base/binaryplatforms.jl

```diff
@@ -1066,20 +1066,12 @@ function select_platform(download_info::Dict, platform::AbstractPlatform = HostP
         return nothing
     end
 
-    # At this point, we may have multiple possibilities.  We now engage a multi-
-    # stage selection algorithm, where we first choose simpler matches over more
-    # complex matches.  We define a simpler match as one that has fewer tags
-    # overall.  As these candidate matches have already been filtered to match
-    # the given platform, the only other tags that exist are ones that are in
-    # addition to the tags declared by the platform. Hence, selecting the
-    # minimum in number of tags is equivalent to selecting the closest match.
-    min_tag_count = minimum(length(tags(p)) for p in ps)
-    filter!(p -> length(tags(p)) == min_tag_count, ps)
-
-    # Now we _still_ may continue to have multiple matches, so we now simply sort
-    # the candidate matches by their triplets and take the last one, so as to
-    # generally choose the latest release (e.g. a `libgfortran5` tarball over a
-    # `libgfortran3` tarball).
+    # At this point, we may have multiple possibilities.  E.g. if, in the future,
+    # Julia can be built without a direct dependency on libgfortran, we may match
+    # multiple tarballs that vary only within their libgfortran ABI.  To narrow it
+    # down, we just sort by triplet, then pick the last one.  This has the effect
+    # of generally choosing the latest release (e.g. a `libgfortran5` tarball
+    # rather than a `libgfortran3` tarball)
     p = last(sort(ps, by = p -> triplet(p)))
```

Comment:
Gonna be a bad surprise when we encounter `libgfortran10`.

Reply:
This secondary sorting only exists to get rid of dict-order selection and use something more deterministic when there is an incomplete specification.  This should never happen with JLLs except in very rare circumstances (the only such circumstance I know of right now is the memory sanitizer builds, which is what these PRs are trying to address).

I've opened a PR to implement a better matching algorithm here: https://github.com/JuliaLang/julia/pull/49502